PyDynamic/model_estimation/fit_filter.py

```diff
@@ -87,92 +86,207 @@ def _fitIIR(
-
-        b, a = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=False)
-        if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-            astab = mapinside(a)
+    as_and_bs = np.empty((mc_runs, Nb + Na + 1))
+    taus = np.zeros((mc_runs,), dtype=int)
+    stab_iters = np.zeros((mc_runs,), dtype=int)
+
+    # Conduct the Monte Carlo runs or in case we did not have uncertainties execute
+    # just once the actual algorithm.
+    for mc_run in range(mc_runs):
+        # Conduct actual fit.
+        b_i, a_i = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=inv)
+
+        # Determine if the computed filter already is stable.
+        unstable = not isstable(b=b_i, a=a_i, ftype="digital")
+
+        # In case the user specified not to check for stability,
+        # we skip the rest of the current Monte Carlo run and inform the user.
+        if justFit:
+            if verbose:
+                sos = np.sum(np.abs((dsp.freqz(b_i, a_i, w)[1] - Hvals) ** 2))
+                print(
+                    f"LSIIR: Fitting "
+                    f"{'' if UHvals is None else f'for MC run {mc_run} '}"
+                    f"finished. No stabilization requested"
+                    f". {warning_unstable if unstable else ''} "
+                    f"Final sum of squares = {sos}"
+                )
+            # Since no stabilization was done, we return tau = 0 regardless of the
+            # desired initial estimate of the time delay provided for the stabilization.
+            taus[mc_run] = 0
+            continue
         else:
-            stable = True
-        run = run + 1
-
-    if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-        print("Caution: The algorithm did NOT result in a stable IIR filter!")
-        print(
-            "Maybe try again with a higher value of tau0 or a higher filter "
-            "order?")
+            # Otherwise we initialize the time delay for the current Monte Carlo run.
+            taus[mc_run] = tau
+
+        # Initialize counter which we use to report about required iteration count.
+        current_stab_iter = 0
+        # Stabilize filter coefficients with a maximum number of iterations.
+        while unstable and current_stab_iter < max_stab_iter:
+            # Compute appropriate time delay for the stabilization of the filter.
+            a_stab = mapinside(a_i)
+            g_1 = grpdelay(b_i, a_i, Fs)[0]
+            g_2 = grpdelay(b_i, a_stab, Fs)[0]
+            taus[mc_run] += np.ceil(np.median(g_2 - g_1))
+
+            # Conduct stabilization step through time delay.
+            b_i, a_i = _fitIIR(Hvals, taus[mc_run], w, E, Na, Nb, inv=inv)
+
+            # Prepare abortion in case filter is stable.
+            unstable = not isstable(b=b_i, a=a_i, ftype="digital")
+            current_stab_iter += 1
+
+        # Finally store filter stacked filter parameters.
+        as_and_bs[mc_run, :] = np.hstack((a_i[1:], b_i))
+        stab_iters[mc_run] = current_stab_iter
```

Comment:
This should be enclosed in a conditional to ensure only stable filters being stored for the computation of the final result.